File: api/routes/pair.go
package routes

import (
	"api/manager"

	"github.com/gofiber/fiber/v2"
)

func PairRoutes(app *fiber.App, sm *manager.SessionManager) {
	api := app.Group("/api")

	api.Post("/instances/:phone/pair", func(c *fiber.Ctx) error {
		phone := c.Params("phone")

		worker, ok := sm.GetWorker(phone)
		if ok {
			status := worker.GetStatus()

			if status == "active" || status == "connected" {
				return c.Status(400).JSON(fiber.Map{"error": "instance already connected"})
			}
		}

		if err := sm.StartInstance(phone, "pairing"); err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "failed to initialize pairing"})
		}

		return c.JSON(fiber.Map{
			"status": "pairing",
			"phone":  phone,
		})
	})
}

File: service/src/routes/pair.rs
use crate::AppState;
use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use serde_json::{Value, json};
use std::sync::Arc;

pub async fn pair_instance(
    Path(phone): Path<String>,
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    {
        let workers = state.sm.workers.read().await;
        if let Some(worker) = workers.get(&phone) {
            if worker.status == "active" || worker.status == "connected" {
                return Err((
                    StatusCode::BAD_REQUEST,
                    Json(json!({"error": "instance already connected"})),
                ));
            }
        }
    }

    state.sm.start_instance(&phone, state.clone()).await;

    Ok(Json(json!({
        "status": "pairing",
        "phone": phone,
    })))
}
